crates/adapters/src/transport/s3/mod.rs

```diff
@@ -0,0 +1,482 @@
+        // objects are being fed to the InputConsumer. We use a bounded channel
+        // to make it so that no more than 8 objects are fetched while waiting
+        // for object processing.
+        let (tx, mut rx) = mpsc::channel(8);
+        tokio::spawn(async move {
+            let mut continuation_token = None;
+            loop {
+                let (objects_to_fetch, next_token): (Vec<String>, Option<String>) = match &config
+                    .read_strategy
+                {
+                    ReadStrategy::Prefix { prefix } => {
+                        let result = client
+                            .get_object_keys(&config.bucket_name, prefix, continuation_token.take())
+                            .await;
+                        if result.is_err() {
+                            error!("Could not fetch object keys {result:?}. Bailing.");
+                            break;
+                        }
+                        result.unwrap()
+                    }
+                    ReadStrategy::SingleKey { key } => (vec![key.clone()], None),
+                };
+                continuation_token = next_token;
+                for key in &objects_to_fetch {
+                    let object = client.get_object(&config.bucket_name, key).await;
+                    tx.send(object).await.expect("Enqueue failed");
+                }
+                if continuation_token.is_none() {
+                    break;
+                }
+            }
+            drop(tx); // We're done. Close the channel.
+        });
+        loop {
+            let state = *receiver.borrow();
+            match state {
+                PipelineState::Paused => {
+                    receiver.changed().await?;
+                }
+                PipelineState::Running => {
+                    tokio::select! {
+                        // If pipeline status has changed, exit the select
+                        _ = receiver.changed() => (),
+                        // Poll the object stream. If `None`, we have processed all objects.
+                        get_obj = rx.recv() => {
+                            if let Some(get_obj) = get_obj {
+                                let mut object = get_obj?;
+                                match object.body.next().await {
+                                    Some(Ok(bytes)) => {
+                                        consumer.input_fragment(&bytes);
+                                    }
+                                    None => break,
+                                    Some(Err(e)) => Err(e)?
+                                }
+                            } else {
+                                // Channel is closed. Exit.
+                                break;
+                            }
+                        }
+                    }
+                }
+                PipelineState::Terminated => break,
```

Comment:
Will this leave the fetching thread running?

Reply:
The worker thread creates the runtime that spawns the fetching task. If the worker thread stops, it should drop the runtime which should stop any spawned tasks running on it as long as they yield (which our fetch task should do so eventually, given the use of await inside it).

Reply:
nice! Would be nice to have a test for this. In fact, it'd be nice to have some unit tests for this adapter.

Reply:
Did you have something else in mind than the tests at the end of the file? 

I'll also add some for the error cases we just discussed.

Reply:
My bad, I actually saw the tests and then forgot that I saw them.